### PR TITLE
Add Azure Client Password as a credentials type

### PR DESCRIPTION
**Pull request by bossinc**
This is for https://github.com/grafana/grafana/pull/89746

Password auth is used for mssql only. This is not a recommended way to auth method, but has been a VOC request. @aangelisc suggested that I put this behind a feature flag, but since we do not make the token in this code, I don't think it is necessary. 

https://learn.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-configure?view=azuresql&tabs=azure-powershell#microsoft-entra-id---password

https://github.com/microsoft/go-mssqldb/blob/main/README.md#azure-active-directory-authentication

### CODE REVIEW

File: aztokenprovider/token_provider.go [hunk 1]
```diff
@@ -106,6 +106,9 @@ func NewAzureAccessTokenProvider(settings *azsettings.AzureSettings, credentials
 			usernameAssertion: tokenEndpoint.UsernameAssertion,
 			tokenRetriever:    tokenRetriever,
 		}, nil
+	case *azcredentials.AzureClientPasswordCredentials:
```
Comment (aangelisc) [inline aztokenprovider/token_provider.go:109]: Is this case required? We could allow this to fall through to the default

File: azcredentials/cloud.go [hunk 2]
```diff
@@ -21,6 +21,8 @@ func GetAzureCloud(settings *azsettings.AzureSettings, credentials AzureCredenti
 		return c.AzureCloud, nil
 	case *AzureClientSecretOboCredentials:
 		return c.ClientSecretCredentials.AzureCloud, nil
+	case *AzureClientPasswordCredentials:
+		return settings.GetDefaultCloud(), nil
```
Comment (aangelisc) [inline azcredentials/cloud.go:25]: I wonder if we should make this configurable - I'm not sure if this works for other cloud types but we can't discount that it does.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,9 @@
 # Change Log
 
+## v2.1.0
+
+- Add Azure Client Password as a credentials type, but there is not a supported azure token provider. You will need to implement the token in your application.
```
Comment (aangelisc) [inline CHANGELOG.md:5]:
```suggestion
- Add Azure Entra Password as a credentials type. This is not a recommended authentication method and is currently only supported for MSSQL. If you require this method the token provider will need to be implemented separately.
```

File: azcredentials/builder.go [hunk 4]
```diff
@@ -124,7 +124,26 @@ func getFromCredentialsObject(credentialsObj map[string]interface{}, secureData
 			},
 		}
 		return credentials, nil
+	case AzureAuthClientPasswordCredentials:
```
Comment (aangelisc) [inline azcredentials/builder.go:127]:
```suggestion
	case AzureEntraPasswordCredentials:
```

File: azcredentials/builder.go [hunk 5]
```diff
@@ -124,7 +124,26 @@ func getFromCredentialsObject(credentialsObj map[string]interface{}, secureData
...
+		clientId, err := maputil.GetString(credentialsObj, "clientId")
+		if err != nil {
+			return nil, err
+		}
+		password, ok := secureData["password"]
+		if !ok {
+			return nil, fmt.Errorf("no password provided")
+		}
 
+		credentials := &AzureClientPasswordCredentials{
```
Comment (aangelisc) [inline azcredentials/builder.go:141]:
```suggestion
		credentials := &AzureEntraPasswordCredentials{
```

File: azcredentials/cloud.go [hunk 6]
```diff
@@ -21,6 +21,8 @@ func GetAzureCloud(settings *azsettings.AzureSettings, credentials AzureCredenti
 		return c.AzureCloud, nil
 	case *AzureClientSecretOboCredentials:
 		return c.ClientSecretCredentials.AzureCloud, nil
+	case *AzureClientPasswordCredentials:
```
Comment (aangelisc) [inline azcredentials/cloud.go:24]:
```suggestion
	case *AzureEntraPasswordCredentials:
```

File: azcredentials/credentials.go [hunk 7]
```diff
@@ -1,11 +1,12 @@
...
-	AzureAuthManagedIdentity     = "msi"
-	AzureAuthWorkloadIdentity    = "workloadidentity"
-	AzureAuthClientSecret        = "clientsecret"
-	AzureAuthClientSecretObo     = "clientsecret-obo"
+	AzureAuthCurrentUserIdentity       = "currentuser"
+	AzureAuthManagedIdentity           = "msi"
+	AzureAuthWorkloadIdentity          = "workloadidentity"
+	AzureAuthClientSecret              = "clientsecret"
+	AzureAuthClientSecretObo           = "clientsecret-obo"
+	AzureAuthClientPasswordCredentials = "ad-password"
```
Comment (aangelisc) [inline azcredentials/credentials.go:9]:
```suggestion
	AzureAuthEntraPasswordCredentials = "ad-password"
```

File: azcredentials/credentials.go [hunk 8]
```diff
@@ -39,6 +40,13 @@ type AzureClientSecretCredentials struct {
 	ClientSecret string
 }
 
+type AzureClientPasswordCredentials struct {
```
Comment (aangelisc) [inline azcredentials/credentials.go:43]:
```suggestion
type AzureEntraPasswordCredentials struct {
```

File: azcredentials/credentials.go [hunk 9]
```diff
@@ -64,3 +72,7 @@ func (credentials *AzureClientSecretCredentials) AzureAuthType() string {
 func (credentials *AzureClientSecretOboCredentials) AzureAuthType() string {
 	return AzureAuthClientSecretObo
 }
+
+func (credentials *AzureClientPasswordCredentials) AzureAuthType() string {
```
Comment (aangelisc) [inline azcredentials/credentials.go:76]:
```suggestion
func (credentials *AzureEntraPasswordCredentials) AzureAuthType() string {
```

File: azcredentials/credentials.go [hunk 10]
```diff
@@ -64,3 +72,7 @@ func (credentials *AzureClientSecretCredentials) AzureAuthType() string {
...
+	return AzureAuthClientPasswordCredentials
```
Comment (aangelisc) [inline azcredentials/credentials.go:77]:
```suggestion
	return AzureAutEntraPasswordCredentials
```

File: azsettings/env.go [hunk 11]
```diff
@@ -27,6 +27,8 @@ const (
 	UserIdentityAssertion                  = "GFAZPL_USER_IDENTITY_ASSERTION"
 	UserIdentityFallbackCredentialsEnabled = "GFAZPL_USER_IDENTITY_FALLBACK_SERVICE_CREDENTIALS_ENABLED"
 
+	ClientPasswordCredentialsEnabled = "GFAZPL_CLIENT_PASSWORD_CREDENTIALS_ENABLED"
```
Comment (aangelisc) [inline azsettings/env.go:30]:
```suggestion
	AzureEntraPasswordCredentialsEnabled = "GFAZPL_AZURE_ENTRA_PASSWORD_CREDENTIALS_ENABLED"
```

File: azsettings/env.go [hunk 12]
```diff
@@ -113,6 +115,14 @@ func ReadFromEnv() (*AzureSettings, error) {
 		azureSettings.UserIdentityFallbackCredentialsEnabled = serviceCredentialsFallback
 	}
 
+	// Client Password Credentials auth
+	if clientPasswordCredentialsEnabled, err := envutil.GetBoolOrDefault(ClientPasswordCredentialsEnabled, false); err != nil {
```
Comment (aangelisc) [inline azsettings/env.go:119]:
```suggestion
	if azureEntraPasswordCredentialsEnabled, err := envutil.GetBoolOrDefault(AzureEntraPasswordCredentialsEnabled, false); err != nil {
```

File: azsettings/env.go [hunk 13]
```diff
@@ -113,6 +115,14 @@ func ReadFromEnv() (*AzureSettings, error) {
...
+		err = fmt.Errorf("invalid Azure configuration: %w", err)
+		return nil, err
+	} else {
+		azureSettings.ClientPasswordCredentialsEnabled = clientPasswordCredentialsEnabled
```
Comment (aangelisc) [inline azsettings/env.go:123]:
```suggestion
		azureSettings.AzureEntraPasswordCredentialsEnabled = azureEntraPasswordCredentialsEnabled
```

File: azsettings/settings.go [hunk 14]
```diff
@@ -25,6 +25,8 @@ type AzureSettings struct {
 
 	CustomCloudList     []*AzureCloudSettings
 	CustomCloudListJSON string
+
+	ClientPasswordCredentialsEnabled bool
```
Comment (aangelisc) [inline azsettings/settings.go:29]:
```suggestion
	AzureEntraPasswordCredentialsEnabled bool
```

File: azsettings/settings.go [hunk 15]
```diff
@@ -126,6 +128,11 @@ func ReadFromContext(ctx context.Context) (*AzureSettings, bool) {
 		}
 	}
 
+	if v := cfg.Get(ClientPasswordCredentialsEnabled); v == strconv.FormatBool(true) {
```
Comment (aangelisc) [inline azsettings/settings.go:131]:
```suggestion
	if v := cfg.Get(AzureEntraPasswordCredentialsEnabled); v == strconv.FormatBool(true) {
```

File: azsettings/settings.go [hunk 16]
```diff
@@ -126,6 +128,11 @@ func ReadFromContext(ctx context.Context) (*AzureSettings, bool) {
...
+		settings.ClientPasswordCredentialsEnabled = true
```
Comment (aangelisc) [inline azsettings/settings.go:132]:
```suggestion
		settings.AzureEntraPasswordCredentialsEnabled = true
```

File: aztokenprovider/token_provider.go (hunk 1 above)
Comment (aangelisc) [inline aztokenprovider/token_provider.go:109]:
```suggestion
	case *azcredentials.AzureEntraPasswordCredentials:
```

Incase we do decide to keep this 😊

File: aztokenprovider/token_provider.go [hunk 17]
```diff
@@ -106,6 +106,9 @@ func NewAzureAccessTokenProvider(settings *azsettings.AzureSettings, credentials
 			usernameAssertion: tokenEndpoint.UsernameAssertion,
 			tokenRetriever:    tokenRetriever,
 		}, nil
+	case *azcredentials.AzureEntraPasswordCredentials:
+		err = errors.New("azure client password credentials are not supported by azure authentication provider")
+		return nil, err
```
Comment (bossinc) [inline aztokenprovider/token_provider.go:111]:
```suggestion
```

**Review by aangelisc — APPROVED** [review]
LGTM, nice work @bossinc!

**Pull request merged**